File: frontend/src/types/agent.ts
// Mirrors backend schemas/agent.py

import type { StrategyConfig, StrategyResponse } from "./strategy";

export type TranslationStatus =
  | "updated_draft"
  | "needs_clarification"
  | "needs_disambiguation"
  | "information"
  | "error";

export interface AppliedOperation {
  operation: string;
  symbol?: string | null;
  description: string;
  // None whenever the user supplied the value directly -- currently
  // only populated for a brand-new rule's capital_allocation
  // defaulting to 5%. See draft_updater.py.
  reasoning?: string | null;
  // Which PortfolioRules field this operation touched, only set for
  // set_portfolio_rule operations (e.g. "total_capital_usd").
  field?: string | null;
}

export interface TranslateRequest {
  conversation_id: string;
  user_id: string;
  message: string;
}

export interface TranslateResponse {
  status: TranslationStatus;
  draft?: StrategyConfig | null;
  applied_operations: AppliedOperation[];
  clarification_message?: string | null;
  disambiguation_message?: string | null;
  disambiguation_candidates: string[];
  error_message?: string | null;
  information_message?: string | null;
  // Always populated regardless of status -- the full composed,
  // multi-Agent narrative for this turn (includes any "Assumptions:
  // ..." disclosures folded in by the backend's ResponseComposer).
  // The forward-looking field; information_message/applied_operations
  // remain the legacy, narrower view of the same turn.
  agent_response: string;
}

export interface ConversationSessionResponse {
  messages: Record<string, unknown>[];
  draft?: StrategyConfig | null;
}

export type ValidationSeverity = "error" | "warning";

export interface ValidationIssue {
  severity: ValidationSeverity;
  symbol?: string | null;
  message: string;
}

export interface ConfirmRequest {
  user_id: string;
  conversation_id: string;
  strategy_id?: string | null;
}

export interface ConfirmRejectedResponse {
  confirmed: false;
  issues: ValidationIssue[];
}

export interface ConfirmAcceptedResponse {
  confirmed: true;
  strategy: StrategyResponse;
  warnings: ValidationIssue[];
}

export type ConfirmResponse = ConfirmAcceptedResponse | ConfirmRejectedResponse;

File: backend/app/schemas/agent.py
"""Request/response schemas for the agent API: translation and confirmation."""

import uuid

from pydantic import BaseModel, ConfigDict

from app.agent.translation.translation_result import AppliedOperation, TranslationStatus
from app.agent.translation.validation import ValidationIssue
from app.schemas.strategy import StrategyConfig, StrategyResponse


class TranslateRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    conversation_id: str
    user_id: uuid.UUID
    message: str


class TranslateResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    status: TranslationStatus
    draft: StrategyConfig | None = None
    applied_operations: list[AppliedOperation] = []
    clarification_message: str | None = None
    disambiguation_message: str | None = None
    disambiguation_candidates: list[str] = []
    error_message: str | None = None
    information_message: str | None = None


class ConfirmRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    user_id: uuid.UUID
    conversation_id: str
    strategy_id: uuid.UUID | None = None


class ConfirmRejectedResponse(BaseModel):
    """Returned (not raised as an error) when validation finds blocking
    issues — nothing was persisted, this is the caller's chance to fix
    the draft and try again."""

    model_config = ConfigDict(extra="forbid")

    confirmed: bool = False
    issues: list[ValidationIssue]


class ConfirmAcceptedResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    confirmed: bool = True
    strategy: StrategyResponse
    warnings: list[ValidationIssue] = []
    
class ConversationSessionResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    messages: list[dict]
    draft: StrategyConfig | None = None